R/join.R

```diff
@@ -13,125 +15,130 @@ new_join_block <- function(
-          observeEvent(
-            sels(),
-            ignoreNULL = FALSE,
-            {
-              shinyjs::toggleState("submit", condition = length(sels()) > 0)
-            }
+          # Initialize join keys module
+          join_keys <- mod_join_keys_server(
+            "join_keys",
+            get_x_cols = get_x_cols,
+            get_y_cols = get_y_cols,
+            initial_keys = by
           )
 
-          # Update by cols when upstream data change
-          observeEvent(
-            cols(),
-            {
-              # Reset sels to NULL if x and y don't have
-              # common columns
-              if (!length(cols())) sels(NULL)
-              updateSelectInput(
-                session,
-                inputId = "by",
-                choices = cols(),
-                selected = sels()
-              )
+          # Update join type when input changes
+          observeEvent(input$type, {
+            if (input$type %in% names(join_types)) {
+              r_join_type(input$type)
             }
-          )
+          })
 
-          observe(
+          # Update join type selector when data changes
+          observe({
             updateSelectInput(
               session,
               inputId = "type",
               choices = join_types,
-              selected = type()
+              selected = r_join_type()
             )
-          )
+          })
 
-          list(
-            expr = eventReactive(
-              input$submit,
+          # Build join expression
+          build_join_expr <- function(join_type, keys) {
+            # Get the dplyr join function
+            join_func <- eval(
               bquote(
-                .(func)(x, y, by = .(cols)),
-                list(
-                  func = eval(
-                    bquote(
-                      as.call(c(as.symbol("::"), quote(dplyr), quote(.(fun)))),
-                      list(fun = as.name(type()))
-                    )
-                  ),
-                  cols = sels()
-                )
+                as.call(c(as.symbol("::"), quote(dplyr), quote(.(fun)))),
+                list(fun = as.name(join_type))
               )
-            ),
+            )
+
+            # Handle different join key formats
+            if (is.character(keys) && length(keys) > 0) {
+              # Simple character vector - natural join
+              bquote(.(func)(x, y, by = .(keys)), list(func = join_func, keys = keys))
+            } else if (is.list(keys) && length(keys) > 0) {
+              # Complex join - handle different name mappings
+              # Convert to dplyr join_by() compatible format
+              join_spec <- if (length(keys) == 1 && is.character(keys[[1]])) {
+                keys[[1]]
+              } else {
+                keys
+              }
+              bquote(.(func)(x, y, by = .(join_spec)), list(func = join_func, join_spec = join_spec))
```

Comment:
Does this allow for column prefixes (`x.`, `y.`) in case we have name clashes? What about non-equi joins? We might need an expression field "escape hatch", no? Also if you're not doing a "natural join", does this go into state and can the constructor be called with this configuration?